src/memray/_memray/hooks.cpp

```diff
@@ -173,7 +175,11 @@ memalign(size_t alignment, size_t size) noexcept
 {
     assert(hooks::memalign);
 
-    void* ret = hooks::memalign(alignment, size);
+    void* ret;
+    {
+        tracking_api::RecursionGuard guard;
```

Comment:
Is this an unrelated change? Or is this necessary because `memalign` in musl calls into a function that we've intercepted?

Reply:
Is exactly because of that. `muslc` doesn't use private symbols in this case. I didn't want to add this everywhere to not impact performance.

Reply:
Hm - would it be better to just move the `RecursionGuard` handling up from `trackAllocationImpl` and `trackDeallocationImpl` into each of the hook functions? 

It would be more consistent, at the cost of a few lines of copy pasted code.

> We only do this (for now) in this call to avoid the extra overhead of the recursion guard in the rest of the allocators (that are more commonly used by far).

> I didn't want to add this everywhere to not impact performance.

What performance are you worried about? This is only hit if the hooks are installed, and in the happy path the call to `valloc` succeeds and we call into `trackAllocation` and it sets the `RecursionGuard` itself. If we hoist the creation of the `RecursionGuard` out of `trackAllocationImpl` and `trackDeallocationImpl` and into the hook functions, then it gets used the same number of times as it used to, so it shouldn't have any performance impact at all. We're already turning it on, we just want to turn it on a bit earlier than we have been.

Or am I missing something?

Reply:
OK. I think this is a bit ugly, and I'd like to see if we can come up with something more elegant, but that can be left for another PR.